### PR TITLE
Pr 19342

### PR DESCRIPTION
**Pull request by petebacondarwin**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: aio/tools/ng-packages-installer.js [hunk 1]
```diff
@@ -106,11 +114,22 @@ class NgPackagesInstaller {
...
     const nodeModulesDir = path.join(rootDir, 'node_modules');
-    const localPackages = this.ngPackages.filter(packageName => this._isLocalPackage(packageName, nodeModulesDir));
+    const packagesDir = path.join(nodeModulesDir, '@angular');
+    let localPackages = [];
+    let localDependencies = [];
+    try {
+      localPackages = shelljs.ls(path.join(packagesDir, '*', NG_LOCAL_FILENAME)).
+        map(path => path.slice(nodeModulesDir.length + 1, - NG_LOCAL_FILENAME.length - 1));
+      localDependencies = shelljs.ls(path.join(nodeModulesDir, '*', NG_LOCAL_FILENAME)).
+          map(path => path.slice(nodeModulesDir.length + 1, - NG_LOCAL_FILENAME.length - 1));
```
Comment (gkalpak) [inline aio/tools/ng-packages-installer.js]: Inconsistent indentation :grin:

File: aio/tools/ng-packages-installer.js [hunk 2]
```diff
@@ -121,7 +140,7 @@ class NgPackagesInstaller {
    * @return {boolean} - True if the package has been overwritten or false otherwise.
    */
   _isLocalPackage(packageName, nodeModulesDir) {
```
Comment (gkalpak) [inline aio/tools/ng-packages-installer.js]: Is this method necessary any more?

File: aio/tools/ng-packages-installer.js [hunk 3]
```diff
@@ -191,14 +210,15 @@ class NgPackagesInstaller {
   }
 
   /**
-   * Remove an installed Angular package from `nodeModulesDir` and replace it with the locally built
+   * Remove an installed Angular package from `targetFolder` and replace it with the locally built
```
Comment (gkalpak) [inline aio/tools/ng-packages-installer.js]: This is no longer `Angular package`, just `package` :smiley:

File: aio/tools/ng-packages-installer.js [hunk 4]
```diff
@@ -191,14 +210,15 @@ class NgPackagesInstaller {
...
    * one. Mark the package by adding an `.ng-local` file in the target directory.
+   * @param {string} sourceFolder   - The folder where we should find the source package
    * @param {string} packageName    - The name of the package to overwrite.
-   * @param {string} nodeModulesDir - The target `node_modules/` directory.
+   * @param {string} targetFolder - The target `node_modules/` directory.
```
Comment (gkalpak) [inline aio/tools/ng-packages-installer.js]: _`node_modules/` directory_ --> _directory_ (or something)

File: aio/tools/ng-packages-installer.js [hunk 5]
```diff
@@ -223,6 +243,21 @@ class NgPackagesInstaller {
     this._log(`Running '${installCmd}' in '${rootDir}'...`);
     shelljs.exec(installCmd, {cwd: rootDir});
   }
+
+  /**
+   * Parse a package's package.json for its peer dependencies, storing them in the `peerDependencies` property.
+   * @param {string} path to a package's package.json
```
Comment (gkalpak) [inline aio/tools/ng-packages-installer.js]: _{string} path to a..._ --> _{string} path - Path to a..._ :grin:

File: aio/tools/ng-packages-installer.js [hunk 6]
```diff
@@ -223,6 +243,21 @@ class NgPackagesInstaller {
...
+   * @returns the path that was passed, to enable chaining
```
Comment (gkalpak) [inline aio/tools/ng-packages-installer.js]: _@returns_ --> _@return_ for consistency 😛

File: aio/tools/ng-packages-installer.js [hunk 7]
```diff
@@ -223,6 +243,21 @@ class NgPackagesInstaller {
...
+   */
+  _capturePeerDependencies(path) {
```
Comment (gkalpak) [inline aio/tools/ng-packages-installer.js]: Order aplhabetically for consistency 😇

**Review by gkalpak — COMMENTED** [review]
As far as I understand, there is still the problem that the copied over peerDependencies could be incompatible with the rest of the aio or examples projects.

But I guess there is only so much we can do about that. If anything, we should strive to keep the dependencies in par.

BTW, direct `dependencies` can also pose problems if they are out of sync.

##
LGTM (except for the missing tests :grin:) as long as Travis is happy too!

File: aio/tools/ng-packages-installer.js (hunk 2 above)
Comment (petebacondarwin) [reply aio/tools/ng-packages-installer.js]: Yes we can drop it

File: aio/tools/ng-packages-installer.js (hunk 7 above)
Comment (petebacondarwin) [reply aio/tools/ng-packages-installer.js]: 😱

**Comment by petebacondarwin** [comment]
This approach is a bit messed up as dependencies have dependencies, etc and so it is not enough even to go for direct (as well as peer) dependencies of the packages :-(

**Comment by gkalpak** [comment]
I played around a bit and here a (semi-verified) plan that might work (along the lines of what @petebacondarwin suggested):

In order to use local Angular packages for the project in directory `foo/` do the following:

1. Copy `foo/package.json` to `foo.package.json.bck`.
2. In `foo/package.json` replace `/"@angular/([^"]+)": "[^"]+"/` (e.g. `"@angular/core": "^4.2.0"`) with `"@angular/$1": "file:./relative/path/to/dist/packages-dist/$1"` (e.g. `"@angular/core": "file:../dist/packages-dist/core"`) - only forpackages that appear in `dist/packages-dist/`.
    (This will take care of direct dependencies.)
3. For each of the packages that were replaced package in step 2, look in `dist/packages-dist/<package>/package.json" for `peerDependencies` and assign the specified versions in the corresponding dependencies (if present) in `foo/package.json`.
    (This will take care of peer dependencies.)
4. Inside `foo/` run `yarn install -- no-lockfile`.
    (This will install the locally built Angular packages and their dependencies/peerDependencies - but without affecting the `yarn.lock` file.)
5. Remove `foo/package.json`.
6. Rename `foo/package.json.bck` to `foo/package.json`.

Note 1: This will still pose problems if the `foo/` project is not compatible with the versions of (peer)Dependencies, but in that case there is nothing the script can do. Manual resolution is necessary.

Note 2: I haven't checked if running `yarn install` in `foo/` will overwrite the updated packages with the one's from `yarn.lock`.

Note 3: Additional "book-keeping" might be needed in order to be able to detect overwritten packages and restore from npm. (In this proves too complicated (given the updated peer dependencies), we can always run `yarn install --force` and re-fetch all packages.)

**Comment by petebacondarwin** [comment]
Closing in favour of #19511

**Pull request closed without merging**